core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerde.java

```diff
@@ -165,161 +136,1033 @@ private static MarkerSet getSerializedPageMarkerSet(Slice serializedPage)
+                int bytesRead = buffer.read(destination, destinationIndex, bytesToRead);
+                if (bytesRead == -1) {
+                    break;
+                }
+                bytesRemaining -= bytesRead;
+                destinationIndex += bytesRead;
+            }
+            return length - bytesRemaining;
+        }
+
+        @Override
+        public void readBytes(byte[] destination, int destinationIndex, int length)
+        {
+            ReadBuffer buffer = buffers[0];
+            int bytesRemaining = length;
+            while (bytesRemaining > 0) {
+                ensureReadable(min(Long.BYTES, bytesRemaining));
+                int bytesToRead = min(bytesRemaining, buffer.available());
+                buffer.readBytes(destination, destinationIndex, bytesToRead);
+                bytesRemaining -= bytesToRead;
+                destinationIndex += bytesToRead;
             }
         }
 
-        private DynamicSliceOutput acquireSliceOutput(int estimatedSize)
+        @Override
+        public void readBytes(Slice destination, int destinationIndex, int length)
         {
-            checkNotClosed();
-            if (sliceOutput != null && sliceOutput.writableBytes() >= estimatedSize) {
-                DynamicSliceOutput result = this.sliceOutput;
-                this.sliceOutput = null;
-                return result;
+            ReadBuffer buffer = buffers[0];
+            int bytesRemaining = length;
+            while (bytesRemaining > 0) {
+                ensureReadable(min(Long.BYTES, bytesRemaining));
+                int bytesToRead = min(bytesRemaining, buffer.available());
+                buffer.readBytes(destination, destinationIndex, bytesToRead);
+                bytesRemaining -= bytesToRead;
+                destinationIndex += bytesToRead;
             }
-            this.sliceOutput = null; // Clear any existing slice output that might be smaller than the request
-            return new DynamicSliceOutput(estimatedSize);
         }
 
-        private void releaseSliceOutput(DynamicSliceOutput sliceOutput)
+        private void ensureReadable(int bytes)
         {
-            if (closed) {
+            if (buffers[0].available() >= bytes) {
                 return;
             }
-            sliceOutput.reset();
-            if (sliceOutput.writableBytes() <= MAX_BUFFER_RETAINED_SIZE) {
-                this.sliceOutput = sliceOutput;
+            decrypt();
+            decompress();
+        }
+
+        private void decrypt()
+        {
+            if (this.encryptionKey.isEmpty()) {
+                return;
             }
+
+            ReadBuffer source = buffers[buffers.length - 1];
+            ReadBuffer sink = buffers[buffers.length - 2];
+            int bytesPreserved = sink.rollOver();
+
+            int encryptedSize = source.readInt();
+            int ivSize = cipher.orElseThrow().getBlockSize();
+            IvParameterSpec iv = new IvParameterSpec(
```

Comment:
Are we ensuring somehow that `source.getSlice()` is at least blockSize long?

Reply:
Oh I think I misread. This is not a temporary space, but sth explicitly written when encrypting.
Still ti does not seem to me we are verifying that `source.getSlice()` looks valid (is long enough)

Reply:
I wanted to avoid additional checks for efficiency reasons. The idea is that if something is off it will fail down the stack (what is less readable, but probably a reasonable price to pay for eliminating extra checks, but maybe I'm over optimizing)